### PR TITLE
add ONNX support for BLOOM

### PR DESCRIPTION
**Pull request by nmntz**
# What does this PR do?

add ONNX support for BLOOM

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

@michaelbenayoun

### CODE REVIEW

**Comment by younesbelkada** [comment]
As you told me offline that the slow tests were passing (under torch1.11.0), looks good to me! Thanks for working on that 🔥

File: src/transformers/models/bloom/configuration_bloom.py [hunk 1]
```diff
@@ -153,3 +159,85 @@ def __init__(
         self.slow_but_exact = slow_but_exact
 
         super().__init__(bos_token_id=bos_token_id, eos_token_id=eos_token_id, **kwargs)
+
+
+# Copied from transformers.models.gpt2.configuration_gpt2.GPT2OnnxConfig
```
Comment (lewtun) [inline src/transformers/models/bloom/configuration_bloom.py:164]: Nice!

File: src/transformers/models/bloom/configuration_bloom.py [hunk 2]
```diff
@@ -153,3 +159,85 @@ def __init__(
...
+    def __init__(
+        self,
+        config: PretrainedConfig,
+        task: str = "default",
+        patching_specs: List[PatchingSpec] = None,
+        use_past: bool = False,
+    ):
+        super().__init__(config, task=task, patching_specs=patching_specs, use_past=use_past)
+        if not getattr(self._config, "pad_token_id", None):
+            # TODO: how to do that better?
```
Comment (lewtun) [inline src/transformers/models/bloom/configuration_bloom.py:175]: Hehe @michaelbenayoun we should fix this sometime :)

File: tests/onnx/test_onnx_v2.py [hunk 3]
```diff
@@ -204,6 +204,7 @@ def test_values_override(self):
 }
 
 PYTORCH_EXPORT_WITH_PAST_MODELS = {
+    ("bloom", "bigscience/bloom-350m"),
```
Comment (lewtun) [inline tests/onnx/test_onnx_v2.py:207]: Have you checked that the slow tests pass for this checkpoint? You can run:

```
RUN_SLOW=1 pytest tests/onnx/test_onnx_v2.py -k "bloom"
```

**Review by lewtun — APPROVED** [review]
Thanks for adding the Bloom config @NouamaneTazi - a very clean PR 💮 !

If the slow tests pass, this PR looks good to me. Let's wait for approval from @LysandreJik or @sgugger before merging this

Edit: I see the CI is now failing for an unrelated issue. I've re-run it, but if it comes up red again, I suggest rebasing on `main` and pushing again

File: src/transformers/models/bloom/configuration_bloom.py [hunk 4]
```diff
@@ -13,7 +13,13 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """ Bloom configuration"""
+from collections import OrderedDict
+from typing import Any, List, Mapping, Optional
+
+from transformers import PreTrainedTokenizer, TensorType, is_torch_available
```
Comment (michaelbenayoun) [inline src/transformers/models/bloom/configuration_bloom.py:19]: I would use strings for type annotations and something like:
```
from typing import TYPE_CHECKING

if TYPE_CHECKING:
   from transformers import ...
```
That way you don't import classes that are not actually being used.

File: src/transformers/models/bloom/modeling_bloom.py [hunk 5]
```diff
@@ -296,8 +293,9 @@ def forward(self, input, mask, max_positions):
             mask = torch.ones(input.shape[0], max_positions, dtype=torch.bool, device=input.device)
 
         mask = mask.to(input.device)
+        seq_ids = torch.arange(max_positions, device=input.device)
         causal_mask = (
-            torch.tril(torch.ones((max_positions, max_positions), dtype=torch.bool))
+            (seq_ids[None, None, :] <= seq_ids[None, :, None])
```
Comment (michaelbenayoun) [inline src/transformers/models/bloom/modeling_bloom.py:298]: Dont know if this is relevant but the original implementation outputs a tensor of rank 2, and your change outputs a tensor of rank 3. Should not be a big deal since we do reshape it afterwards but just wanted to point this out.

**Review by michaelbenayoun — APPROVED** [review]
Great, thanks for handling this!

**Comment by younesbelkada** [comment]
@sgugger I do not think this will hurt performances in terms of logits since slow tests are passing, but might hurt indeed the inference time performance for large and/or batched sequences.. We need to benchmark that though to be sure

File: tests/onnx/test_onnx_v2.py (hunk 3 above)
Comment (NouamaneTazi) [reply tests/onnx/test_onnx_v2.py:207]: Thank you for reminding me. All tests are passing now 🙂

File: src/transformers/models/bloom/modeling_bloom.py [hunk 6]
```diff
@@ -78,17 +78,14 @@ def split_tensor_along_last_dim(tensor, num_partitions, contiguous_split_chunks=
...
-        attention_mask_bool = (1 - attention_mask).bool()
+    attention_mask_bool = ~attention_mask.bool()
 
     query_length, key_length, n_heads = attention_scores.size(2), attention_scores.size(3), attention_scores.size(1)
-    padded_causal_mask = (
-        attention_mask_bool[:, None, key_length - query_length : key_length, None]
-        + ~causal_mask[:, :, key_length - query_length : key_length, :key_length]
-    ).bool()
-    padded_causal_mask = padded_causal_mask + attention_mask_bool[:, None, None, :key_length].bool()
+    padded_causal_mask = torch.logical_or(
```
Comment (michaelbenayoun) [inline src/transformers/models/bloom/modeling_bloom.py:84]: This looks good!

File: src/transformers/models/bloom/modeling_bloom.py (hunk 5 above)
Comment (michaelbenayoun) [reply src/transformers/models/bloom/modeling_bloom.py:298]: Do we keep it like that?

File: src/transformers/models/bloom/configuration_bloom.py [hunk 7]
```diff
@@ -153,3 +163,85 @@ def __init__(
...
+        # Need to add the past_keys
+        if self.use_past:
+            if not is_torch_available():
+                raise ValueError("Cannot generate dummy past_keys inputs without PyTorch installed.")
+            else:
+                import torch
+
+                batch, seqlen = common_inputs["input_ids"].shape
+                # Not using the same length for past_key_values
+                past_key_values_length = seqlen + 2
```
Comment (michaelbenayoun) [inline src/transformers/models/bloom/configuration_bloom.py:225]: I think you mention in one commit that the past shape is not the same as the one in `gpt-2`, are the dynamic axes left unchanged though? Asking because otherwise it can break you `inputs` property.

**Review by michaelbenayoun — APPROVED** [review]
LGTM

**Comment by michaelbenayoun** [comment]
Also make sure all the tests pass before merging.

File: src/transformers/models/bloom/configuration_bloom.py (hunk 7 above)
Comment (NouamaneTazi) [reply src/transformers/models/bloom/configuration_bloom.py:225]: Yes, the aim of that commit was to fix the `past_key_values` length.
I did make sure that the inputs work properly now, and the tests are passing

**Comment by NouamaneTazi** [comment]
All tests for `tests/onnx/test_onnx_v2.py -k "bloom"` and `tests/models/bloom` are passing.
Here are the ones that are skipped (which is fine according to @younesbelkada)
```
================================================================================= short test summary info =================================================================================
SKIPPED [1] tests/test_modeling_common.py:2006: test is PT+FLAX test
SKIPPED [1] tests/test_modeling_common.py:1934: test is PT+FLAX test
SKIPPED [1] tests/test_modeling_common.py:1758: test is PT+TF test
SKIPPED [1] tests/test_tokenization_common.py:1960: This test is only for slow tokenizers
SKIPPED [1] tests/test_tokenization_common.py:2189: test is PT+TF test
================================================================= 159 passed, 5 skipped, 35 warnings in 449.50s (0:07:29)
```

**Comment by sgugger** [comment]
There is a difference between a copy in BLOOM and the original in GPT-2 which is why the CI is failing. Make sure to run `make fic-copies` or remove the `Copied from`.

File: src/transformers/models/bloom/configuration_bloom.py [hunk 8]
```diff
@@ -165,7 +165,7 @@ def __init__(
         super().__init__(bos_token_id=bos_token_id, eos_token_id=eos_token_id, **kwargs)
 
 
-# Copied from transformers.models.gpt2.configuration_gpt2.GPT2OnnxConfig with modified `generate_dummy_inputs`
+# Taken from transformers.models.gpt2.configuration_gpt2.GPT2OnnxConfig with modified `generate_dummy_inputs`
```
Comment (sgugger) [inline src/transformers/models/bloom/configuration_bloom.py:168]: The comment can be removed entirely then, there is no point leaving it
```suggestion
```

File: src/transformers/models/bloom/modeling_bloom.py (hunk 5 above)
Comment (NouamaneTazi) [reply src/transformers/models/bloom/modeling_bloom.py]: Fixed :) Thank you for the notice 🤗

**Review by sgugger — APPROVED** [review]
Thanks a lot!
